### PR TITLE
Persistent GuildPrefix Caching/Saving [Redis]

### LINKED ISSUE #1: Storing prefixes for the servers.
I think there should be option too choose a option where we want to store our data, like mysql, moongose, sqlite, yaml, json etc.

### PR DESCRIPTION
**Pull request by oleggtro**
I just completed the redis integration for persistent prefix storing as wished in #1 ... Its tested, I may implement loading the settings from the settings.yaml and not hardcoding it...

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -31,6 +31,22 @@
                     <target>1.8</target>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>3.2.4</version>
+                <configuration>
```
Comment (Kaktushose) [inline pom.xml:38]: not needed, should be removed

File: pom.xml [hunk 2]
```diff
@@ -60,6 +76,16 @@
             <artifactId>slf4j-api</artifactId>
             <version>1.7.30</version>
         </dependency>
+        <dependency>
+            <groupId>redis.clients</groupId>
+            <artifactId>jedis</artifactId>
+            <version>3.2.0</version>
+        </dependency>
+        <dependency>
```
Comment (Kaktushose) [inline pom.xml]: shade is only needed as a plugin, not a dependency

File: src/main/java/com/github/kaktushose/jda/commands/api/EmbedFactory.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public MessageEmbed getSpecificHelpEmbed(@Nonnull CommandCallable commandCallabl
      * @param event    the corresponding {@code GuildMessageReceivedEvent}
      * @return the MessageEmbed to send
      */
-    public MessageEmbed getCommandNotFoundEmbed(@Nonnull CommandSettings settings, @Nonnull GuildMessageReceivedEvent event) {
+    public MessageEmbed getCommandNotFoundEmbed(CommandSettings settings, GuildMessageReceivedEvent event) {
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/api/EmbedFactory.java]: removal of `Nonnull` annotation intended?

File: src/main/java/com/github/kaktushose/jda/commands/entities/Parameter.java [hunk 4]
```diff
@@ -29,6 +29,7 @@
      * @param defaultValue  the default value for the parameter, only used if isOptional is true
      * @param parameterType the {@link ParameterType} of the parameter
      */
+
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/Parameter.java]: line should be removed

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java [hunk 5]
```diff
@@ -125,9 +146,10 @@ public static JDACommands start(@Nonnull ShardManager shardManager, @Nullable St
      * @param botMentionPrefix whether to allow a bot mention to be a valid prefix or not
      * @return a {@link JDACommands} instance that has started the initialization process
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java:151]: missing javadoc

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java [hunk 6]
```diff
@@ -141,9 +163,10 @@ public static JDACommands start(@Nonnull JDA jda, @Nullable String prefix, boole
...
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java:172]: missing javadoc

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java [hunk 7]
```diff
@@ -160,7 +183,25 @@ public JDACommandsBuilder setSettings(@Nonnull CommandSettings settings) {
...
+        this.isRedisEnabled = isRedisEnabled;
+        this.redisHost = redisHost;
+        this.redisPort = redisPort;
+        this.redisDatabase = redisDatabase;
+        return this;
+    }
+
+
+    /**
+     * Changes the {@link EventParser} used to parse incoming {@code 1MessageReceivedEvent}s.
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java]: typo

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java [hunk 8]
```diff
@@ -244,7 +285,7 @@ public JDACommands build() {
         try {
             settings = YamlLoader.load(JDACommands.class.getClassLoader().getResource("settings.yaml"));
             log.debug("Found a settings file. Maybe overriding given runtime values");
-        } catch (IOException | NullPointerException ignore) {
+        } catch (IOException | NullPointerException | ClassNotFoundException ignore) {
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java]: `ClassNotFoundException` has nothing to do with a missing settings file. No reason to ignore it

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java [hunk 9]
```diff
@@ -35,6 +36,10 @@
     private ArgumentParser argumentParser;
     private EmbedFactory embedFactory;
     private HelpMessageSender helpMessageSender;
+    private boolean isRedisEnabled;
+    private String redisHost;
+    private Integer redisPort;
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java:41]: usage of int wrapper class intended?

File: src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java [hunk 10]
```diff
@@ -123,6 +124,10 @@ public void onGuildMessageReceived(@Nonnull GuildMessageReceivedEvent event) {
         Optional<CommandCallable> command = commandMapper.findCommand(commands, input, settings.isIgnoreLabelCase());
         if (!command.isPresent()) {
             log.debug("No command for input {} found", Arrays.toString(input));
+
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:127]: no reason to null check these objects

File: src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java [hunk 11]
```diff
@@ -55,7 +56,7 @@ public CommandDispatcher(Object jda,
         this.eventParser = eventParser;
         this.commandMapper = commandMapper;
         this.argumentParser = argumentParser;
-        this.embedFactory = embedFactory;
+        this.embedFactory = new EmbedFactory();
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:59]: This means that own embed factories cannot be used. This should be undo

File: src/main/java/com/github/kaktushose/jda/commands/internal/YamlLoader.java [hunk 12]
```diff
@@ -3,6 +3,9 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
 import com.github.kaktushose.jda.commands.entities.CommandSettings;
+import com.github.kaktushose.jda.commands.entities.JDACommands;
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/internal/YamlLoader.java]: unused imports

File: src/main/java/com/github/kaktushose/jda/commands/internal/YamlLoader.java [hunk 13]
```diff
@@ -22,15 +25,14 @@
...
-
     /**
      * Loads a yaml file and creates a {@link CommandSettings} object from it.
      *
      * @param url The URL where the yaml file is located
      * @return The {@link CommandSettings} object
      * @throws IOException If an I/O error occurs
      */
-    public static CommandSettings load(URL url) throws IOException {
+    public static CommandSettings load(URL url) throws IOException, ClassNotFoundException {
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/internal/YamlLoader.java]: same as for JDACommandsBuilder. `ClassNotFoundException` has nothing to do with settings files

File: src/main/java/com/github/kaktushose/jda/commands/internal/RedisSettingsHolder.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        this.redisDatabase = redisDatabase;
+    }
+
+    /**
+     * This gets the boolean, which specifies if jedis should or shouldn't be enabled
+     *
+     * @return bool if redis should be enabled or shouldn't
+     */
+
+    public boolean getRedisEnabled() {
```
Comment (Kaktushose) [inline src/main/java/com/github/kaktushose/jda/commands/internal/RedisSettingsHolder.java]: should be named `isRedisEnabled`

**Review by Kaktushose — CHANGES_REQUESTED** [review]
Basically, the changes are good, but many minor and some major problems still need to be addressed.
In addition to the comments, I noticed the following things:
- remove the `dependecy-reduced-pom` 
- please use dependency injection instead of singleton

File: src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java (hunk 10 above)
Comment (oleggtro) [reply src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:127]: Was only used for debugging when discovering the embedFactory  NullPointerException. Just missed it, while cleaning the code.

File: src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java (hunk 11 above)
Comment (oleggtro) [reply src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:59]: For whatever reason embedFactory is null when using `this.embedFactory = embedFactory;``.
I fixed it by just creating a new one.

File: src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java (hunk 9 above)
Comment (oleggtro) [reply src/main/java/com/github/kaktushose/jda/commands/entities/JDACommandsBuilder.java:41]: Usage was intended due to the possibility for the user not to use redis. I did not want to set the port to something random. As of now I am working on specifying redis options in the settings.yml, therefore it won't be needed in the future anymore.

File: src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java (hunk 11 above)
Comment (Kaktushose) [reply src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:59]: Well but your fix disables other functionality. However the NPE is related to a bug/missing statement I've just found. In the JDACommandsBuilder constructor using the ShardManager, the EmbedFactory is not instantiated, resulting in a NPE

Comment (oleggtro) [reply src/main/java/com/github/kaktushose/jda/commands/internal/CommandDispatcher.java:59]: Do you want me to correct it, or do you want to do it yourself?

**Pull request merged**